pallas-miniprotocols/src/txsubmission/server.rs

```diff
@@ -0,0 +1,144 @@
+use pallas_codec::Fragment;
+use pallas_multiplexer::agents::{Channel, ChannelBuffer};
+
+use super::protocol::{Message, State, TxBody, TxId, TxIdAndSize, Error, Blocking, TxCount};
+
+pub enum Reply {
+    TxIds(Vec<TxIdAndSize>),
+    Txs(Vec<TxBody>),
+    Done,
+}
+
+pub struct Server<H>(State, ChannelBuffer<H>)
+where
+    H: Channel,
+    Message: Fragment;
+
+impl<H> Server<H>
+where
+    H: Channel,
+    Message: Fragment,
+{
+    pub fn new(channel: H) -> Self {
+        Self(State::Init, ChannelBuffer::new(channel))
+    }
+
+    pub fn state(&self) -> &State {
+        &self.0
+    }
+
+    pub fn is_done(&self) -> bool {
+        self.0 == State::Done
+    }
+
+    // NOTE(pi): as of this writing, the network spec has a typo; this is the correct behavior
+    fn has_agency(&self) -> bool {
+        match self.state() {
+            State::Idle => true,
+            _ => false,
+        }
+    }
+
+    fn assert_agency_is_ours(&self) -> Result<(), Error> {
+        if !self.has_agency() {
+            Err(Error::AgencyIsTheirs)
+        } else {
+            Ok(())
+        }
+    }
+
+    fn assert_agency_is_theirs(&self) -> Result<(), Error> {
+        if self.has_agency() {
+            Err(Error::AgencyIsOurs)
+        } else {
+            Ok(())
+        }
+    }
+
+    /// As a server in a specific state, am I allowed to send this message?
+    fn assert_outbound_state(&self, msg: &Message) -> Result<(), Error> {
+        match (&self.0, msg) {
+            (State::Idle, Message::RequestTxIds(..)) => Ok(()),
+            (State::Idle, Message::RequestTxs(..)) => Ok(()),
+            _ => Err(Error::InvalidInbound),
+        }
+    }
+
+    /// As a server in a specific state, am I allowed to receive this message?
+    fn assert_inbound_state(&self, msg: &Message) -> Result<(), Error> {
+        match (&self.0, msg) {
+            (State::Init, Message::Init) => Ok(()),
+            (State::TxIdsBlocking, Message::ReplyTxIds(..)) => Ok(()),
+            (State::TxIdsBlocking, Message::Done) => Ok(()),
+            (State::TxIdsNonBlocking, Message::ReplyTxIds(..)) => Ok(()),
+            (State::Txs, Message::ReplyTxs(..)) => Ok(()),
+            _ => Err(Error::InvalidOutbound),
+        }
+    }
+
+    pub fn send_message(&mut self, msg: &Message) -> Result<(), Error> {
+        self.assert_agency_is_ours()?;
+        self.assert_outbound_state(msg)?;
+        self.1.send_msg_chunks(msg).map_err(Error::ChannelError)?;
+
+        Ok(())
+    }
+
+    pub fn recv_message(&mut self) -> Result<Message, Error> {
+        self.assert_agency_is_theirs()?;
+        let msg = self.1.recv_full_msg().map_err(Error::ChannelError)?;
+        self.assert_inbound_state(&msg)?;
+
+        Ok(msg)
+    }
+
+    pub fn wait_for_init(&mut self) -> Result<(), Error> {
+        if self.0 != State::Init {
```

Comment:
Was wondering if I should put some kind of loop here, in case we're waiting for a connection or something, but kept it very minimal;

If there's interest, I could make a `recv_init` that does the dumb thing, and a `wait_for_init` that loops until someone connects and sends the init message.

Reply:
well, the `recv_message` blocks until there's something to read from the channel, the "wait" is imposed to the consumer, even if they don't want it.

Ideally, every blocking call should provide an optional timeout, so that the control flow logic can escape the blocking call if desired. This is an improvement we should make at the multiplexer level, there should be a way to read from a channel with a timeout.

Reply:
Right but if someone hasn't even connected yet this errors immediately; or is there a construct at the multiplexer / channel level that lets me wait for a connection?

Reply:
I see. I think that our server agent should be created as a result of a successful bearer connection (aka: tcp or unix socket connection).

I imagine the root process waiting for connections on a specific port, this would be a blocking call that is handled by the OS (something like `TcpListener::bind("127.0.0.1:80")`). Upon each connection, the process would create a new agent passing a channel connected to our new bearer (inbound socket stream). By this point in time, the calls to `recv_message` should not fail, but rather block waiting for data.

Summarizing, I think that the initial loop / block is out-of-scope for this crate. I say we create a new example crate (different PR) that shows how to connect our agent to a tcp listener. What do you think?

Reply:
Makes sense to me; there's nothing in the type system that currently enforces this, though; i.e. I would expect something guiding me to "hey, you opened a connection to the relay, not the other way around, so you can't use this socket/channel to create a server"; see the integration test I added for example.

Is it worth adding a ClientChannel and ServerChannel type to distinguish this?